Svc/FramingProtocol/docs/sdd.md

```diff
@@ -1,29 +1,314 @@
+
+The abstract class `FramingProtocol` provides a protected member `m_interface`.
+This member is a pointer, initially null.
+After the `setup` method of `FramingProtocol` is called, it
+points to a concrete instance of `FramingProtocolInterface`.
+
+Your implementation of `frame` should do the following:
+
+1. Use `m_interface->allocate` to allocate a buffer to hold the framed data.
+
+1. Frame the data into the buffer allocated in step 1.
+
+1. Use `m_interface->send` to send the buffer.
+
+### 3.2. Deframing
+
+To implement a deframing protocol, do the following:
+
+1. Implement the abstract class `DeframingProtocolInterface`.
 
-    virtual void route(Fw::Buffer& data) = 0;
+1. Use the implementation in step 1 to implement the abstract class
+`DeframingProtocol`.
+
+#### 3.2.1. Implementing `DeframingProtocolInterface`
+
+`DeframingProtocolInterface` defines helper methods for deframing data.
+Typically these methods are implemented by an F Prime component (e.g., `Svc::Deframer`),
+because they require port invocations.
+The component `Svc::Deframer` provides an implementation of `DeframingProtocolInterface`
+that you can use.
+It does this by inheriting from `DeframingProtocolInterface` and implementing
+its abstract methods.
+
+To implement `DeframingProtocolInterface`, you must implement the following
+pure virtual methods:
+
+```c++
+virtual Fw::Buffer allocate(const U32 size) = 0;
+
+virtual void route(Fw::Buffer& data) = 0;
+```
+
+The method `allocate` should allocate memory, as described in
+Section 3.1.1.
+
+The method `route` should send (route) the data stored in the buffer.
+A typical implementation invokes either an `Fw::Com` port (e.g., for sending
+commands) or a `Fw::BufferSend` port (e.g., for sending file packets).
+
+#### 3.2.2. Implementing `DeframingProtocol`
+
+`DeframingProtocol` defines the operation of deframing a packet.
+To implement `DeframingProtocol`, you must implement the following pure
+virtual method:
+
+```c++
+virtual DeframingStatus deframe(Types::CircularBuffer& buffer, U32& needed) = 0;
 ```
 
-Diagram view of FramingProtocol:
+This method is called with the following arguments:
+
+* `buffer`: A circular buffer holding the data to deframe.
+
+* `needed`: A reference for returning the number of bytes needed
+for deframing.
+
+`deframe` returns a value of type `DeframingStatus` indicating what happened.
+
+The abstract class `DeframingProtocol` provides a protected member `m_interface`.
+It operates as described in Section 3.1.2.
+
+Your implementation of `deframe` should do the following:
+
+1. Determine how many bytes are needed for deframing, peeking into
+the circular buffer if necessary (e.g., to read a length).
+
+1. If that many bytes are available in the circular buffer:
+
+   1. Use `m_interface->allocate` to allocate an `Fw::Buffer`
+to hold the deframed data.
+
+   1. Deframe the data into the allocated buffer.
+
+   1. Use `m_interface->route` to send the buffer.
+
+1. Return status.
```

Comment:
Need to ensure that the user sets the "needed" value correctly too.

Reply:
Fixed.